CSharp/README.md

```diff
@@ -314,6 +315,84 @@ public async Task<int> GetLatestPositionAsync()
 }
 ```
 
+## Testing Code
+
+We all know testing code is not built the same way that regular code is, because its purpose is different. This is why we have special guidelines for this type of code.
+
+### Test Class naming: {TargetTestClass}Tests
+
+```csharp
+// Bad
+namespace MyProject.UnitTests
+{
+    [TestClass]
+    public class MyService
+    { }
+}
+
+// Good
+namespace MyProject.UnitTests
+{
+    [TestClass]
+    public class MyServiceTests
+    { }
+}
+```
+
+### Test Class Location: keep a correspondence with the code being tested
+
+Use the same folder structure / namespace structure as the code that's being tested. This makes it easier to find test classes for a particular class and classes being tested by a particular test class.
+
+### Method naming: {TargetTestMethod}_{Expectation}[_When{Condition}]
+
+The name needs to express:
+
+- **What** is being tested
+- **What** should happen.
+- Under **which** conditions.
+
+What is being tested will usually be a particular method. (These are unit tests, after all.) For integration tests, there is more freedom in choosing this portion, but should still be a good name indicating the scenario being tested.
+
+What should happen should be concise and to the point. Avoid _should_ or _must_ or expressions of rule/desire. The test itself is that validation, there's no need to specify it.

```

Comment:
I understand that there is not need to specify it, but I think that it is useful to encourage programmers to try to synthesize in a _human sentence_ what the test do, in a _human_ way.

Also, we could have a convention and avoid also the _when_, but it is less _human_ yet.

Also, Microsoft _cool teams_ are using `_should_`, see https://github.com/aspnet/Home/wiki/Engineering-guidelines#unit-test-method-naming